src/lib/io/sol-iio.c

```diff
@@ -700,12 +741,33 @@ sol_iio_open(int device_id, const struct sol_iio_config *config)
         return NULL;
     }
 #endif
+    SOL_LIST_FOREACH (&iio_open_list, it) {
+        entry = SOL_LIST_GET_CONTAINER(it, struct iio_opened_device, list);
+        if (device_id == entry->device->device_id) {
+            if ((config->buffer_size == entry->config->buffer_size)
+                && (config->sol_iio_reader_cb == entry->config->sol_iio_reader_cb)
+                && (config->sampling_frequency == entry->config->sampling_frequency)) {
+                entry->refcount++;
+                device = entry->device;
+                reader_cb_data_elem = calloc(1, sizeof(struct reader_cb_data));
+                SOL_NULL_CHECK(reader_cb_data_elem, NULL);
+                reader_cb_data_elem->data = config->data;
+                sol_list_append(&device->reader_cb_data_list, &reader_cb_data_elem->list);
+                return device;
+            } else {
+                SOL_ERR("device%d is already open, but could not reuse it with different config",
+                    device_id);
+                return NULL;
+            }
+        }
+    }
 
     device = calloc(1, sizeof(struct sol_iio_device));
     SOL_NULL_CHECK(device, NULL);
     sol_ptr_vector_init(&device->channels);
     device->buffer = SOL_BUFFER_INIT_EMPTY;
-    device->fd = -1;
+    device->reader_cb_data_list.next = &device->reader_cb_data_list;
+    device->reader_cb_data_list.prev = &device->reader_cb_data_list;
```

Comment:
I didn't get this change. Why remove `device->fd` initialisation? Why are you making list point to list?